subprojects/core/src/main/java/org/gradle/api/internal/artifacts/DependencyResolutionServices.java

```diff
@@ -15,18 +15,25 @@
  */
 package org.gradle.api.internal.artifacts;
 
+import org.gradle.api.artifacts.ConfigurationContainer;
 import org.gradle.api.artifacts.dsl.DependencyHandler;
 import org.gradle.api.artifacts.dsl.DependencyLockingHandler;
 import org.gradle.api.artifacts.dsl.RepositoryHandler;
 import org.gradle.api.attributes.AttributesSchema;
-import org.gradle.api.internal.artifacts.configurations.RoleBasedConfigurationContainerInternal;
 import org.gradle.api.internal.attributes.ImmutableAttributesFactory;
 import org.gradle.api.model.ObjectFactory;
 
+/**
+ * Provides access to services required for dependency resolution.
+ * <p>
+ * Note that changes to this type, even seemingly safe ones such as narrowing the return types, can
+ * cause problems for IDEs (the IDE tests should fail upon such changes, alerting us to
+ * this problem).  Thus, this internal API should be treated as semi-public.
```

Comment:
I am really not a fan of this wording here. What does it mean to treat APIs as semi-public? Instead we should be specific about where this internal API is being used, preferably linking to the code. This way in the future we can make informed decisions on what we can and can't change here. 

Seemingly this documentation belongs on `getConfigurationContainer` instead of the class. I don't think we need to avoid changes to this whole class just cause they call one method on it.

Reply:
My goal was just to put something here to give us a heads up.  I'm not sure which methods are or aren't used, or where the IDE uses them, and I don't think we can/should link to the failing tests in the IDE tests repo.  

Semi-public is some vaguely defined location between public API and internal, with all the drawbacks of both. 😁 It's a "Here Be Dragons".

I think it'd be great if someone with more knowledge improved on this, and if we started using some annotations like: `@UsedByIDE("IntelliJ")`, `@UsedByPlugin("some.plugin")` throughout our internal types.  I'm going to suggest adding these in the appropriate forum.  

This is better than nothing though, and would have prevented this type of problem, so I'm going to leave it as is.

Reply:
I don't think this is in "widespread use" across the IDE. This is a one-off use in one class in the IDE. There are no other references to `DependencyResolutionServices` except in https://github.com/JetBrains/intellij-community/blob/04f12153ef33e31f9e5eed0212da2858c2b85642/plugins/gradle/tooling-extension-impl/src/org/jetbrains/plugins/gradle/tooling/builder/VersionCatalogsModelBuilder.java#L12

To treat this whole class as semi-public just cause there is one usage of it in the IDE seems heavy handed